### ISSUE TITLE
Avatars in git blame for everyone

### ISSUE THREAD

**Issue by WeetHet** [zed-industries/zed#11043]
### Check for existing issues

- [X] Completed

### Describe the feature

Currently, only GitHub avatars are supported in git blame. Other hostings should be supported as well.

## Avatar support
- [x] GitHub
- [ ] GitLab (main instance)
- [ ] Gitee
- [ ] Bitbucket
- [ ] Sourcehut
- [x] Codeberg (#10991)
- [ ] Self-hosted instances of GitLab
- [ ] Self-hosted instances of Gitea
- [ ] Self-hosted instances of Forjejo
- [ ] Self-hosted instances of Sourcehut

 Anything else with an API

**Comment by mrnugget** [comment]
> * Should this be an API for extensions?

I think for now it's fine to keep it in there, since it's similar to the permalinks behavior we have.

> * Should we make a `Commit`/`Author`/`User`/... struct for each hosting separately or extract them?
>   For now I've extracted the `Author` struct as it's the only one that is built by hand and not deserialised from the response.

I don't think we need more than a URL right now. Once we need more, we can add more.

**Comment by WeetHet** [comment]
> > * Should this be an API for extensions?
> 
> I think for now it's fine to keep it in there, since it's similar to the permalinks behavior we have.
> 
> > * Should we make a `Commit`/`Author`/`User`/... struct for each hosting separately or extract them?
> >   For now I've extracted the `Author` struct as it's the only one that is built by hand and not deserialised from the response.
> 
> I don't think we need more than a URL right now. Once we need more, we can add more.

Okay, I'll undraft the PR and we'll go from there

**Comment by d1y** [comment]
The `Gitee` host need `access_token`  before you can send request fetch avatar
So the user may need to fill it in manually, this is a setting item 🤔

<img width="420" src="https://github.com/zed-industries/zed/assets/45585937/41b0af7a-75ba-48e8-8ced-1a501a197bf2" />


> API: https://gitee.com/api/v5/swagger#/getV5ReposOwnerRepoCommitsSha

```ts
function createURL(owner: string, repo: string, sha: string, token: string) {
  return `https://gitee.com/api/v5/repos/${owner}/${repo}/commits/${sha}?access_token=${token}`
}

const token = Bun.env["ACCESS_TOKEN"]!

const url = createURL('libkitten', 'zed', 'f72cf2afe34f77bf498454cd5486aa2c31b820d6', token)

console.log(url)

// export ACCESS_TOKEN=$$$
// bun run run.ts | xargs curl | jless
```

Avatar is `author.avatar_url`

**Comment by amtoaer** [comment]
@d1y I just gave it a try, and it turns out that access token is not necessary, just like the Swagger documentation indicates:

<img width="2238" height="1082" alt="Image" src="https://github.com/user-attachments/assets/c0258e4e-234b-4a4c-8ef2-593f9295e455" />

For setting the image size, they use `!avatar${size}` on their website, but I found that `?width=${size}` is also supported.

**Comment by amtoaer** [comment]
The support for **self-hosted GitHub Enterprise** instances has been implemented in #26482, and **Sourcehut** seems to completely lack avatar support. Therefore, after #41813 is merged, the only remaining unimplemented avatar feature on the list should be for **Bitbucket**.

However, I cannot successfully register on their cloud instance, as it continuously throws the error: `Something went wrong on our end. Please try again.`, making development and debugging difficult.

---

Furthermore, I believe there is a critical related issue that needs discussion: The current provider implementations can correctly fetch public repositories, but my understanding is that fetching private repositories would certainly require users to provide their own credentials.

We are facing two main challenges:

1.  **Token Scoping:** While some request code supports reading environment variables (like `GITHUB_TOKEN`) as a workaround, this variable is tied solely to the Provider type. Even within GitHub, **Self-Hosted** and **Public Instances** are two different providers, and having them share a single token is clearly incorrect.

2.  **Unreliable Detection:** The current self-hosted detection relies only on keyword matching. As the existing comment states:

    > TODO: detecting self hosted instances by checking whether "xxx" is in the url or not is not very reliable

    For example, if I deploy a Gitea instance with the URL `git.ea.org`, the existing code will fail to recognize it correctly.

My proposed solution is to allow users to specify the association between a URL and a provider in a configuration file, along with their personal token. The configuration could look like this:

```json
{
  "provider": [
    { "url": "https://my.git.ea.org", "provider": "GITEA", "token": "xxx" },
    { "url": "https://my.git.lab.com", "provider": "GITLAB", "token": "xxx" }
  ]
}
```

This approach would solve both the token scoping and instance detection problems simultaneously. However, storing tokens directly in a configuration file does not seem like best practice. Perhaps the best method would be to store them directly in the **System Keychain**, similar to how `AI_KEYs` are handled.

I would like to hear everyone's thoughts on this proposal. Regardless, I believe this architectural change is significantly more complex than implementing the git blame avatar feature and should perhaps be tracked as a separate issue.

**Comment by WeetHet** [comment]
> My proposed solution is to allow users to specify the association between a URL and a provider in a configuration file, along with their personal token. The configuration could look like this:
> 
> {
>   "provider": [
>     { "url": "https://my.git.ea.org", "provider": "GITEA", "token": "xxx" },
>     { "url": "https://my.git.lab.com", "provider": "GITLAB", "token": "xxx" }
>   ]
> }

Yeah, this was my idea of how this should be done too, I've just never gotten around to it. I'm not so sure about storing the tokens in plain text in the settings, but setting up which host corresponds to which provider should definitely be implemented like this

**Comment by amtoaer** [comment]
Ah, after digging into the code, I found that the provider's setting already exists. The only remaining problem is how to supply different tokens for the public instance versus the self-hosted instance.
